### PR TITLE
categories

### PR DESCRIPTION
**Pull request by jayesh-parmar**

### CODE REVIEW

File: app/Http/Controllers/CategoryController.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+{
+    public function index()
+    {
+        $categories = Category::with('children')->whereNull('parent_id')->get();
+        return view('admin.pages.category.index', compact('categories'));
+    }
+
+    public function add()
+    {
+        $categories = Category::select('id', 'name', 'description', 'parent_id')->get();
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: why this required?

File: app/Http/Controllers/CategoryController.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    public function store(CategoryRequest $request)
+    {
+        Category::create($request->all());
+
+        return redirect()->route('categories.index')->with('success', 'Added successfully ');
+    }
+    public function edit(string $catId)
+    {
+        $categories = Category::select('id', 'name', 'description', 'parent_id')->get();
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: why $categories fetch?

File: app/Http/Controllers/CategoryController.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        return view('admin.pages.category.form', compact('categories'));
+    }
+
+    public function store(CategoryRequest $request)
+    {
+        Category::create($request->all());
+
+        return redirect()->route('categories.index')->with('success', 'Added successfully ');
+    }
+    public function edit(string $catId)
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]:
```suggestion
    public function edit(string $categoryId)
```

File: app/Http/Controllers/CategoryController.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $categories = Category::select('id', 'name', 'description', 'parent_id')->get();
+        $category = Category::select('id', 'name', 'description', 'parent_id')->find($catId) ;
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: `parent_id` required?

File: app/Http/Controllers/CategoryController.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        
+        return view('admin.pages.category.form', compact('category','categories'));
+    }
+    public function update(CategoryRequest $request, Category $category)
+    {
+        $category->update($request->all());
+
+        return redirect()->route('categories.index')->with('success', 'Category Updated successfully.');
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]:
```suggestion
        return redirect()->route('categories.index')->with('success', 'Category updated successfully.');
```

File: app/Http/Controllers/CategoryController.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public function destroy($id)
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]:
```suggestion
    public function destroy(string $categoryId)
```

File: app/Http/Controllers/CategoryController.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $category = Category::find($id);
+
+        $category->children()->delete();
+
+        $category->delete();
+
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: if parent have child category and trying to delete, it should not be delete.

File: app/Http/Requests/CategoryRequest.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array|string>
+     */
+    public function rules(): array
+    {
+        return [
+            'name' => ['required', 'string', 'max:255'],
+            'description' => ['nullable', 'string'],
+            'parent_id' => ['nullable', 'exists:categories,id'],
```
Comment (rajan-parmar) [inline app/Http/Requests/CategoryRequest.php]: parent_id is not required

File: app/Http/Requests/CategoryRequest.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array|string>
+     */
+    public function rules(): array
+    {
+        return [
+            'name' => ['required', 'string', 'max:255'],
```
Comment (rajan-parmar) [inline app/Http/Requests/CategoryRequest.php]: name should be unique

File: app/Models/Category.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    protected $fillable = [
+        'name',
+        'description',
+        'parent_id'
+    ];
+
+    public function parent()
+    {
+        return $this->belongsTo(Category::class, 'parent_id');
+    }
```
Comment (rajan-parmar) [inline app/Models/Category.php]: it's not required

File: resources/views/admin/pages/category/form.blade.php [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+        <select id="parent_id" name="parent_id" class="form-control">
+            <option value="">Select Parent Category</option>
+            @foreach($categories as $cat)
+            <option value="{{ $cat->id }}" @if(isset($category)) {{ $category->parent_id == $cat->id ? 'selected' : '' }} @endif>{{ $cat->name }}</option>
+            @endforeach
+        </select>
+    </div><br>
+    <div class="form-group">
+        <label for="description">Description</label>
+        <textarea id="description" name="description" value="{{ old('name', $category->description ?? '') }}" class="form-control">@if (isset($category)){{$category->description }}@endif</textarea>
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/form.blade.php]: image


avi rite banavanu che.

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/CategoryController.php (hunk 1 above)
Comment (jayesh-parmar) [reply app/Http/Controllers/CategoryController.php]: parent category list show karva mate

File: app/Http/Controllers/CategoryController.php (hunk 2 above)
Comment (jayesh-parmar) [reply app/Http/Controllers/CategoryController.php]: form ma parent category list show karva mate

File: app/Http/Controllers/CategoryController.php (hunk 4 above)
Comment (jayesh-parmar) [reply app/Http/Controllers/CategoryController.php]: sir update ma list print karava jarur pade atle lidhi che

File: resources/views/admin/pages/category/form.blade.php (hunk 11 above)
Comment (jayesh-parmar) [reply resources/views/admin/pages/category/form.blade.php]: Screenshot (21)
 sir am dekhay che chal che ?

File: app/Http/Controllers/CategoryController.php (hunk 1 above)
Comment (rajan-parmar) [reply app/Http/Controllers/CategoryController.php]: jarur nathi. tamare dropdown ma select nathi karvani

File: app/Http/Controllers/CategoryController.php (hunk 2 above)
Comment (rajan-parmar) [reply app/Http/Controllers/CategoryController.php]: jarur nathi

File: app/Http/Controllers/CategoryController.php (hunk 4 above)
Comment (rajan-parmar) [reply app/Http/Controllers/CategoryController.php]: jarur nathi. tamare record display j karvo che.

File: app/Http/Controllers/CategoryController.php [hunk 12]
```diff
@@ -24,28 +24,30 @@ public function store(CategoryRequest $request)
...
 
-    public function destroy($id)
+    public function destroy(string $categoryId)
     {
-        $category = Category::find($id);
+         $category = Category::select('id')->find($categoryId);
 
+        if ($category->children->isNotEmpty()) {
+            return redirect()->route('categories.index')->with('error', 'Cannot delete category with child categories.');
+        }
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: query is not proper.

message is: `This category cannot be deleted as there are one or more child categories attached.`

File: app/Http/Requests/CategoryRequest.php [hunk 13]
```diff
@@ -22,9 +23,8 @@ public function authorize(): bool
     public function rules(): array
     {
         return [
-            'name' => ['required', 'string', 'max:255'],
+            'name' => ['required', 'string', 'max:255', Rule::unique('categories')->ignore($this->category, 'name')],
```
Comment (rajan-parmar) [inline app/Http/Requests/CategoryRequest.php]:
```suggestion
            'name' => ['required', 'string', 'max:255', Rule::unique('categories')->ignore($this->category)],
```

File: resources/views/admin/pages/category/index.blade.php [hunk 14]
```diff
@@ -1,8 +1,19 @@
...
+<div>
+    {{ session('success') }}
+</div>
+@endif
+
+@if (session('error'))
+<div style="color: red;">
+    {{ session('error') }}
+</div>
+@endif
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/index.blade.php]: set in layout file.

File: resources/views/admin/pages/category/subcategory.blade.php [hunk 15]
```diff
@@ -1,11 +1,16 @@
...
-    <form action="{{ route('categories.destroy', $category->id) }}" method="POST" class="d-inline">
-        @csrf
-        <button type="submit" class="btn btn-danger" onclick="return confirm('Are you sure you want to delete this category?')">Delete</button>
-    </form>
+    <div style="display: flex; align-items: center; background-color: #f0f0f0;">
+        <p style="margin-right: 10px;">({{ $category->children->count() }}) </p>
+        <h5 style="margin-right: 10px;">{{ $category->name }}</h5>
+        <div style="margin-right: 10px;">{{ $category->description }}</div>
+        <a href="{{ route('categories.edit', $category->id) }}" class="btn btn-warning" style="margin-right: 10px;text-decoration: none;">Edit</a>
+        <a href="{{route('categories.create')}}" style="margin-right: 10px; text-decoration: none;">Add Category</a>
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/subcategory.blade.php]: adding child category then you have to pass parent category id

File: resources/views/admin/pages/category/subcategory.blade.php [hunk 16]
```diff
@@ -1,11 +1,16 @@
...
+        <form action="{{ route('categories.destroy', $category->id) }}" method="POST">
+            @csrf
+            <button type="submit" style="border: none; color: red; background: none; cursor: pointer;" onclick="return confirm('Are you sure you want to delete this category?')">Delete</button>
+        </form>
+    </div>
+    <hr>
     @if ($category->children->count() > 0)
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/subcategory.blade.php]: condition not required

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/CategoryController.php [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+            $parentCategoryId = Category::find($parentId);
+
+            if (!$parentCategoryId) {
+                return redirect()->back()->with('error', 'Parent category not found.');  
+            }
+            return view('admin.pages.category.form', compact('parentCategoryId'));
+        }
+        else {
+            return view('admin.pages.category.form');
+        }
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]:
```suggestion
        public function add($parentId = null)
{
    return view('admin.pages.category.form', compact('parentId'));
}
```

File: app/Http/Controllers/CategoryController.php [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+    public function store(CategoryRequest $request)
+    {
+        Category::create($request->validated());
+
+        return redirect()->route('categories.index')->with('success', 'Category added successfully ');
+    }
+    public function edit(string $categoryId)
+    {
+        $category = Category::select('id', 'name', 'description', 'parent_id')->find($categoryId);
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]:
```suggestion
        $category = Category::select('id', 'name', 'description')->find($categoryId);
```

File: app/Http/Controllers/CategoryController.php [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        else {
+            return view('admin.pages.category.form');
+        }
+    }
+
+    public function store(CategoryRequest $request)
+    {
+        Category::create($request->validated());
+
+        return redirect()->route('categories.index')->with('success', 'Category added successfully ');
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]:
```suggestion
        return redirect()->route('categories.index')->with('success', 'Category added successfully.');
```

File: app/Http/Controllers/CategoryController.php [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    public function update(CategoryRequest $request, Category $category)
+    {
+        $category->update($request->validated());
+
+        return redirect()->route('categories.index')->with('success', 'Category updated successfully.');
+    }
+
+    public function destroy(string $categoryId)
+    {
+         $category = Category::find($categoryId);
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: just check that $categoryId is in `parent_id` or not.

File: app/Http/Controllers/CategoryController.php [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+        if ($category->children->isNotEmpty()) {
+            return redirect()->route('categories.index')->with('error', 'This category cannot be deleted as there are one or more child categories attached.');
+        }
+        $category->children()->delete();
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: why remove children?

File: app/Http/Requests/CategoryRequest.php [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+class CategoryRequest extends FormRequest
+{
+    /**
+     * Determine if the user is authorized to make this request.
+     */
+    public function authorize(): bool
+    {
+        return true;
+    }
```
Comment (rajan-parmar) [inline app/Http/Requests/CategoryRequest.php]: not required

File: app/Http/Requests/CategoryRequest.php [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array|string>
+     */
+    public function rules(): array
+    {
+        return [
+            'name' => ['required', 'string', 'max:255', Rule::unique('categories')->ignore($this->category)],
+            'description' => ['nullable', 'string'],
+            'parent_id' => ['nullable', 'string'],
```
Comment (rajan-parmar) [inline app/Http/Requests/CategoryRequest.php]: use exists & max rule

File: resources/views/admin/pages/category/form.blade.php [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+            @if($errors->has('name'))
+            <span class="text-xs text-red-600 dark:text-red-400">{{ $errors->first('name') }}</span>
+            @endif
+        </label>
+        @if(isset($parentCategoryId))
+        <input type="hidden" name="parent_id" value="{{ $parentCategoryId->id }}">
+        <label class="block text-sm">
+            <span class="text-gray-700 dark:text-gray-400">Parent Category Name</span>
+            <input class="block w-full mt-1 text-sm dark:text-gray-300 dark:border-gray-600 dark:bg-gray-700 form-select focus:border-purple-400 focus:outline-none focus:shadow-outline-purple dark:focus:shadow-outline-gray" type="text" name="parent_name" value="{{ $parentCategoryId->name }}" disabled />
+        </label>
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/form.blade.php]: keep single input for name.

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/CategoryController.php (hunk 21 above)
Comment (jayesh-parmar) [reply app/Http/Controllers/CategoryController.php]: just check karva mate

File: app/Http/Controllers/CategoryController.php [hunk 25]
```diff
@@ -49,12 +39,12 @@ public function update(CategoryRequest $request, Category $category)
 
     public function destroy(string $categoryId)
     {
-         $category = Category::find($categoryId);
+        $category = Category::where('id', $categoryId)->first();
 
         if ($category->children->isNotEmpty()) {
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: tamare te check karvanu che k $categoryId parent_id ma che k nahi. jo hoy to means te koi child no parent che. 
`if ($category->children->isNotEmpty()) {` not required

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

File: resources/views/admin/pages/category/subcategory.blade.php [hunk 26]
```diff
@@ -0,0 +1,19 @@
+<li class="category-item">
+    <div style="display: flex; align-items: center; background-color: #f0f0f0;">
+        <p style="margin-right: 10px;">({{ $category->children->count() }}) </p>
+        <h5 style="margin-right: 10px;">{{ $category->name }}</h5>
+        <div style="margin-right: 10px;">{{ $category->description }}</div>
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/subcategory.blade.php]: description is not requried

File: resources/views/admin/pages/category/subcategory.blade.php [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+        <a href="{{ route('categories.edit', $category->id) }}" class="btn btn-warning" style="margin-right: 10px;text-decoration: none;">Edit</a>
+        <a href="{{ route('categories.create', ['parentId' => $category->id]) }}" style="margin-right: 10px; text-decoration: none;">Add Child Category</a>
+        <form action="{{ route('categories.destroy', $category->id) }}" method="POST">
+            @csrf
+            <button type="submit" style="border: none; color: red; background: none; cursor: pointer;" onclick="return confirm('Are you sure you want to delete this category?')">Delete</button>
+        </form>
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/subcategory.blade.php]: a sequence like this:
add child
edit
delete

File: resources/views/admin/pages/category/subcategory.blade.php [hunk 28]
```diff
@@ -0,0 +1,19 @@
+<li class="category-item">
+    <div style="display: flex; align-items: center; background-color: #f0f0f0;">
+        <p style="margin-right: 10px;">({{ $category->children->count() }}) </p>
+        <h5 style="margin-right: 10px;">{{ $category->name }}</h5>
```
Comment (rajan-parmar) [inline resources/views/admin/pages/category/subcategory.blade.php:4]:
```suggestion
        <h5 style="margin-right: 10px;">{{ $category->name }} ({{ $category->children->count() }})</h5>
```

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

**Review by rajan-parmar — APPROVED** [review]

**Pull request merged**